File: implementations/java/src/main/java/io/rcl/Lexer.java
package io.rcl;

import java.util.ArrayList;
import java.util.List;

final class Lexer {
  static List<Token> lex(String src) {
    List<Token> out = new ArrayList<>();
    int i = 0, line = 1, col = 1, n = src.length();
    while (i < n) {
      char ch = src.charAt(i);
      if (ch == ' ' || ch == '\t' || ch == '\r') { i++; col++; continue; }
      if (ch == '\n') { i++; line++; col = 1; continue; }
      if (ch == '#') { while (i < n && src.charAt(i) != '\n') { i++; col++; } continue; }
      if (ch == '\'') throw err("single-quoted string usage", line, col);
      if (ch == '/' && i + 1 < n && src.charAt(i + 1) == '/') throw err("unexpected character", line, col);
      if (idStart(ch)) {
        int s = i, c0 = col;
        while (i < n && idPart(src.charAt(i))) { i++; col++; }
        String v = src.substring(s, i);
        out.add(new Token(v.equals("do") ? TokenKind.DO : v.equals("end") ? TokenKind.END : TokenKind.IDENT, v, line, c0));
        continue;
      }
      if (ch == '-' || Character.isDigit(ch)) {
        int s = i, c0 = col;
        if (ch == '-') {
          if (i + 1 >= n || !Character.isDigit(src.charAt(i + 1))) throw err("unexpected character", line, col);
          i++; col++;
        }
        while (i < n && Character.isDigit(src.charAt(i))) { i++; col++; }
        if (i < n && src.charAt(i) == '.') {
          if (i + 1 >= n || !Character.isDigit(src.charAt(i + 1))) throw err("unexpected character", line, col);
          i++; col++;
          while (i < n && Character.isDigit(src.charAt(i))) { i++; col++; }
        }
        out.add(new Token(TokenKind.NUMBER, src.substring(s, i), line, c0));
        continue;
      }
      if (ch == '"') {
        int c0 = col;
        StringBuilder sb = new StringBuilder();
        i++; col++;
        while (i < n && src.charAt(i) != '"') {
          char x = src.charAt(i);
          if (x == '\n') throw err("unterminated string", line, c0);
          if (x == '\\') {
            i++; col++;
            if (i >= n) throw err("unterminated string", line, c0);
            char e = src.charAt(i);
            if (e == '"') sb.append('"');
            else if (e == '\\') sb.append('\\');
            else if (e == 'n') sb.append('\n');
            else if (e == 't') sb.append('\t');
            else throw err("invalid escape", line, col);
            i++; col++;
            continue;
          }
          sb.append(x); i++; col++;
        }
        if (i >= n) throw err("unterminated string", line, c0);
        i++; col++;
        out.add(new Token(TokenKind.STRING, sb.toString(), line, c0));
        continue;
      }
      TokenKind k = switch (ch) {
        case '=' -> TokenKind.EQ;
        case ',' -> TokenKind.COMMA;
        case '.' -> TokenKind.DOT;
        case '[' -> TokenKind.LBR;
        case ']' -> TokenKind.RBR;
        default -> null;
      };
      if (k == null) throw err("unexpected character", line, col);
      out.add(new Token(k, String.valueOf(ch), line, col));
      i++; col++;
    }
    out.add(new Token(TokenKind.EOF, "", line, col));
    return out;
  }

  private static boolean idStart(char ch) { return Character.isLetter(ch) || ch == '_'; }
  private static boolean idPart(char ch) { return Character.isLetterOrDigit(ch) || ch == '_'; }
  static ParseError err(String msg, int line, int col) { return new ParseError(msg, line, col); }
}

File: implementations/kotlin/src/main/kotlin/io/rcl/Lexer.kt
package io.rcl

enum class TokenType { Identifier, String, Number, Equal, Comma, Dot, Do, End, LBracket, RBracket, EOF }
data class Token(val type: TokenType, val value: String, val line: Int, val column: Int)

data class LexerState(val pos: Int, val line: Int, val column: Int)

class Lexer(private val input: String) {
  private var pos = 0
  private var line = 1
  private var column = 1

  fun snapshot() = LexerState(pos, line, column)
  fun restore(s: LexerState) { pos = s.pos; line = s.line; column = s.column }

  fun nextToken(): Token {
    skip()
    if (eof()) return Token(TokenType.EOF, "", line, column)
    val line0 = line
    val col0 = column
    val ch = cur()
    return when {
      ch == '=' -> one(TokenType.Equal, "=", line0, col0)
      ch == ',' -> one(TokenType.Comma, ",", line0, col0)
      ch == '.' -> one(TokenType.Dot, ".", line0, col0)
      ch == '[' -> one(TokenType.LBracket, "[", line0, col0)
      ch == ']' -> one(TokenType.RBracket, "]", line0, col0)
      ch == '"' -> readString(line0, col0)
      ch.isDigit() || (ch == '-' && peek().isDigit()) -> readNumber(line0, col0)
      ch.isLetter() || ch == '_' -> readIdentifier(line0, col0)
      else -> throw ParseError("unexpected character", line0, col0)
    }
  }

  private fun readString(line0: Int, col0: Int): Token {
    adv()
    val out = StringBuilder()
    while (!eof() && cur() != '"') {
      if (cur() == '\\') {
        adv()
        if (eof()) throw ParseError("unterminated escape", line0, col0)
        out.append(
          when (cur()) {
            '"' -> '"'; 'n' -> '\n'; 't' -> '\t'; '\\' -> '\\'
            else -> throw ParseError("invalid escape", line, column)
          }
        )
        adv()
      } else { out.append(cur()); adv() }
    }
    if (eof()) throw ParseError("unterminated string", line0, col0)
    adv()
    return Token(TokenType.String, out.toString(), line0, col0)
  }

  private fun readNumber(line0: Int, col0: Int): Token {
    val out = StringBuilder()
    if (cur() == '-') { out.append('-'); adv() }
    while (!eof() && cur().isDigit()) { out.append(cur()); adv() }
    if (!eof() && cur() == '.' && peek().isDigit()) {
      out.append('.'); adv(); while (!eof() && cur().isDigit()) { out.append(cur()); adv() }
    }
    return Token(TokenType.Number, out.toString(), line0, col0)
  }

  private fun readIdentifier(line0: Int, col0: Int): Token {
    val out = StringBuilder()
    while (!eof() && (cur().isLetterOrDigit() || cur() == '_')) { out.append(cur()); adv() }
    val value = out.toString()
    val type = when (value) { "do" -> TokenType.Do; "end" -> TokenType.End; else -> TokenType.Identifier }
    return Token(type, value, line0, col0)
  }

  private fun skip() {
    while (!eof()) {
      when {
        cur().isWhitespace() -> adv()
        cur() == '#' -> while (!eof() && cur() != '\n') adv()
        else -> return
      }
    }
  }

  private fun one(type: TokenType, value: String, line0: Int, col0: Int): Token { adv(); return Token(type, value, line0, col0) }
  private fun adv() { if (eof()) return; if (cur() == '\n') { line++; column = 1 } else column++; pos++ }
  private fun cur() = input[pos]
  private fun peek() = input.getOrElse(pos + 1) { '\u0000' }
  private fun eof() = pos >= input.length
}
